### ISSUE TITLE
Built in locking (feature)

### ISSUE THREAD

**Issue by jamespharaoh** [zbackup/zbackup#134]
I propose that ZBackup should include a locking mechanism, to prevent more than one operation from updating the repository at the same time.

Since the purpose of zbackup is to deduplicate, and this cannot be done by concurrent processes (in the current design, at least), I never want more than one process to backup to the same repository at once.

I use `flock` in my shell scripts which invoke zbackup, but I think this is an essential feature that should be added to the core. It is especially important when removing items from the repository, eg garbage collection or other potential modifications which may be implemented in the future.

I propose to add a lock flag to the respository info, to set it on by default, but of course it should be able to be configured, both when creating a repository and afterwards. This would be implemented by creating a "lock" file in the repository, and peforming an exclusive flock on it, whenever making changes to the repository.

This also allows us to clean up the "tmp" directory, which often gets filled with junk from failed backup processes, since we can be fairly confident that noone else is writing there at the present time.

Command line options would also enable this to be bypassed one a one-off basis.

Comments please.

**Comment by mknjc** [comment]
Flock is problematic over network filesystems, I think a lot of users have the zbackup dir mounted over NFS, CIFS or other network filesystems. We could add a lockfile but if a zbackup process crashes, the system dies, ... it need user interaction because how should zbackup make sure its not running on a different system.

**Comment by jamespharaoh** [comment]
I am using flock over NFS and it works perfectly. I think, in general, this works for NFS4, but not for older versions.

I believe that if flock is not supported it will simply fail, so a simple test could be made when creating a repository.

I feel that nothing is lost by having this option available, and a lot is gained by making shell scripts, and manual usage, much simpler. I can't imagine I'm the only person who makes an effort to avoid multiple concurrent backups to increase deduplication.

For example, I have a large number of LXC containers. If I perform an upgrade on them, they'll all have new copies of the same files. Every day when this happens, if the backups run in parallel, I will lose the ability to deduplicate this data.

**Comment by mknjc** [comment]
So your idea is to flock optimistic and if it fails, it fails? Zbackup should handle the cause of two instances writing into to same repo quite well and the only thing you would loose is the better deduping.

**Comment by jamespharaoh** [comment]
Yes, pretty much that, and yes, the safety that is provided already is part of the reason I see this as a fairly reasonable change.

Whether it would be enabled by default would be a different question, but I'd certainly like to set a flag in the repo's metadata so that it only has to be enabled once.

Different options could allow for an error in the case that the flock fails entirely, etc.

**Comment by timvaillancourt** [comment]
I'm also implementing locking around ZBackup and I think this is a good idea. If there is concern for NFS-based deploys maybe this should just be an optional flag for the zbackup command with default: disabled. Would that work for everyone?

**Comment by siretty** [comment]
You could run `zbackup` using `runlock` from the cronutils package in Debian; place the lockfile (flag `-f`) in the mount and you should be good to go.
